docs-internal/archetype-v2.md

```diff
@@ -0,0 +1,883 @@
+
+## Flow
+
+The flow is the graph of inputs.
+
+### Input
+
+The flow is formed by way of nesting `<input>` elements. `<input>` requires a `name` attribute that must be unique
+ among siblings.
+
+An input can be of different types:
+- `type="option"`: an option (`true` or `false`) ; opt-in by default but can be declared as opt-out (`default="false"`)
+- `type="select"`: pick and choose ; single optional choice by default, but can be required (`required="true"`) and or multiple (`multiple="true"`)
+- `type="text"`: text value ; may have a default value (`placeholder="my-default-value"`)
+
+An input can also be declared as hidden. Hidden inputs are not exposed in UIs. The only way to set a value of hidden
+ inputs is via preset (`<context>`, CLI options, URL query params etc).
+
+Example of a selection:
+```xml
+<input name="tracing"
+       type="select"
+       label="Select a tracing provider">
+
+    <option value="zipkin" label="Zipkin" />
+    <option value="jaeger" label="Jaeger" />
+</input>
+```
+
+Example of an option:
+```xml
+<input name="kubernetes"
+       type="option"
+       label="Kubernetes Support"
+       prompt="Do you want support for Kubernetes"/>
+```
+
+Example of a text value:
+```xml
+<input name="name"
+       type="text"
+       label="Project name"
+       placeholder="myproject"/>
+```
+
+### Step
+
+A step represents a UX pane, or a window that contains a certain set of inputs.
+Steps are required by default and can be made optional using the optional attribute (`optional="true`).
+
+```xml
+<step label="Application Type">
+    <!-- ... -->
+</step>
+```
+
+### Optional Steps
+
+Optional steps must have inputs with defaults. An optional step with non default inputs is invalid and should be
+ validated as an error.
+
+Customization of features can be modeled using an optional step declared as the last child of the step to be
+ customized. The order is the declaration order, the enclosing steps must declare the optional step used for
+ customization carefully.
+
+E.g.
+- `se/se.xml` executes `se/hello-world/hello-world-se.xml` AND THEN executes `common/customize-project.xml`
+- `se/hello-world/hello-world-se.xml` executes `se/hello-world/customize-hello-world.xml` as the very last step.
+
+The declaration order for the two customization steps is first `customize-hello-world` and then `customize-project`.
+
+When interpreting the flow, the next steps are always computed. If the next steps are all optional, it is possible to
+ skip them and generate the project.
+
+An optional step requires no action other than continuing to the next step (unless this is the very last step).
+
+## Flow Context
+
+The flow context is a tree of resolved inputs that represents the choices. The tree nodes represent choices that
+ are a pair of an input path its resolved value.
+
+The context may be populated with various mechanisms:
+- `<context>` directive
+- query parameters
+- CLI option
+
```

Comment:
Insert something here about fully specifying the presets such that no choices remain.